internal/mode/static/nginx/config/upstreams_test.go

```diff
@@ -273,13 +354,107 @@ func TestCreateUpstream(t *testing.T) {
 			},
 			msg: "endpoint ipv6",
 		},
+		{
+			stateUpstream: dataplane.Upstream{
+				Name: "single upstreamSettingsPolicy",
+				Endpoints: []resolver.Endpoint{
+					{
+						Address: "10.0.0.1",
+						Port:    80,
+					},
+				},
+				Policies: []policies.Policy{
+					&ngfAPI.UpstreamSettingsPolicy{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      "usp",
+							Namespace: "test",
+						},
+						Spec: ngfAPI.UpstreamSettingsPolicySpec{
+							ZoneSize: helpers.GetPointer[ngfAPI.Size]("2m"),
+							KeepAlive: helpers.GetPointer(ngfAPI.UpstreamKeepAlive{
+								Connections: helpers.GetPointer(int32(1)),
+								Requests:    helpers.GetPointer(int32(1)),
+								Time:        helpers.GetPointer[ngfAPI.Duration]("5s"),
+								Timeout:     helpers.GetPointer[ngfAPI.Duration]("10s"),
+							}),
+						},
+					},
+				},
+			},
+			expectedUpstream: http.Upstream{
+				Name:     "single upstreamSettingsPolicy",
+				ZoneSize: "2m",
+				Servers: []http.UpstreamServer{
+					{
+						Address: "10.0.0.1:80",
+					},
+				},
+				KeepAliveConnections: 1,
+				KeepAliveRequests:    1,
+				KeepAliveTime:        "5s",
+				KeepAliveTimeout:     "10s",
+			},
+			msg: "single upstreamSettingsPolicy",
+		},
+		{
+			stateUpstream: dataplane.Upstream{
+				Name: "multiple upstreamSettingsPolicies",
+				Endpoints: []resolver.Endpoint{
+					{
+						Address: "10.0.0.1",
+						Port:    80,
+					},
+				},
+				Policies: []policies.Policy{
+					&ngfAPI.UpstreamSettingsPolicy{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      "usp1",
+							Namespace: "test",
+						},
+						Spec: ngfAPI.UpstreamSettingsPolicySpec{
+							ZoneSize: helpers.GetPointer[ngfAPI.Size]("2m"),
+							KeepAlive: helpers.GetPointer(ngfAPI.UpstreamKeepAlive{
+								Time:    helpers.GetPointer[ngfAPI.Duration]("5s"),
+								Timeout: helpers.GetPointer[ngfAPI.Duration]("10s"),
+							}),
+						},
+					},
+					&ngfAPI.UpstreamSettingsPolicy{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      "usp2",
+							Namespace: "test",
+						},
+						Spec: ngfAPI.UpstreamSettingsPolicySpec{
+							KeepAlive: helpers.GetPointer(ngfAPI.UpstreamKeepAlive{
+								Connections: helpers.GetPointer(int32(1)),
+								Requests:    helpers.GetPointer(int32(1)),
+							}),
+						},
+					},
+				},
+			},
+			expectedUpstream: http.Upstream{
+				Name:     "multiple upstreamSettingsPolicies",
+				ZoneSize: "2m",
+				Servers: []http.UpstreamServer{
+					{
+						Address: "10.0.0.1:80",
+					},
+				},
+				KeepAliveConnections: 1,
+				KeepAliveRequests:    1,
+				KeepAliveTime:        "5s",
+				KeepAliveTimeout:     "10s",
+			},
+			msg: "multiple upstreamSettingsPolicies",
```

Comment:
I think we are missing the test case where the zone size is not set by any policy.

Reply:
Added a test case for an empty UpstreamSettingsPolicy, does this satisfy this case? Or would you like me to create a test case with an UpstreamSettingsPolicy which sets keepalive settings and for me to check the ZoneSIze?

Reply:
I think both tests cases are necessary 🙂